### PR TITLE
File related opcodes fixes. More tests.

### PR DESCRIPTION
**Pull request by MiranDMC**
Fixes https://github.com/cleolibrary/CLEO5/issues/58

### CODE REVIEW

File: cleo_plugins/FileSystemOperations/FileSystemOperations.cpp [hunk 1]
```diff
@@ -315,25 +317,66 @@ class FileSystemOperations
...
+            }
+
+            // local variable array
+            case DT_LVAR_ARRAY:
+            case DT_LVAR_TEXTLABEL:
+            case DT_LVAR_TEXTLABEL_ARRAY:
+            case DT_LVAR_STRING_ARRAY:
+            {
+                auto localsBeg = (char*)thread + offsetof(CLEO::CRunningScript, LocalVar);
+                auto localsEnd = localsBeg + sizeof(CLEO::CRunningScript::LocalVar) + sizeof(CLEO::CRunningScript::Timers); // timers are in fact just two extra local variables
```
Comment (x87) [inline cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: that's not the original behavior. just bring back original implementation from CLEO4.

If you need some safety checks,

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: When user provide long string variable as target he expects to read into it, no into 5 other next variables if file data happens to be longer.
The fact you can write 40 characters into 4@v does not mean any other opcode will be able to read it, as these types are not based on null termination.

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: Maybe it makes sense anyway. That data can be accessed with GetVarPointer. FileRead also does not care about target size.

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: If I provide an int value or int variable, or element of an int array, copy {size} bytes from file to that address.
If I provide a string variable (or element of a string array), copy {size} bytes to the address of that variable.

Don't limit users on what they can and can not do.

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: I may use entire local variables space (32 variables) as a temp buffer to read chunks of data from file (128 bytes), modify them and write back. I don't need any intervention with data from runtime, just raw copy.

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp:351]: Only thing we should do is to ensure null-terminator at the {size} position, if fgets does not do it already.
And gate negative size/invalid pointers.

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: Updated.
fgets reads size-1 by default.

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: I'm thinking of a use case like:
```
0@v = "0123456789ABCDEF"
read_string_from_file file_handle to 0@v size 16
```
will the string end with \0?

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: Yes, 15 characters and terminator is written.

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]:
```
{$CLEO .cs}
nop

0@ = open_file "cleo\.cleo.log" {mode} "r" // tested opcode
1@v = "0123456789ABCDEF"
read_string_from_file 0@ to 1@v size 16

0AD1: print_formatted_now fmt "%s" time 200000 arg 1@v

terminate_this_custom_script
```

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: I mean, can you update test suite?

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: Ehhh, ok :]

Comment (x87) [reply cleo_plugins/FileSystemOperations/FileSystemOperations.cpp]: probably makes sense to gradually built out comprehensive test suite, one fix at a time

**Review by x87 — APPROVED** [review]

File: cleo_plugins/FileSystemOperations/FileUtils.cpp [hunk 2]
```diff
@@ -84,20 +85,57 @@ bool File::flush(DWORD handle)
...
+			else
+			{
+				modeValid = false;
+				break; // invalid character
+			}
+
+			i++;
+		}
+
+		if (!modeBin) modeParsed[i] = 'b'; // always open as binary mode, as text mode does not behave as expected in multiple scenarios
```
Comment (x87) [inline cleo_plugins/FileSystemOperations/FileUtils.cpp:116]: can you explain these scenarios?

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileUtils.cpp:116]: Generally text mode is not well documented and many stream related functions has undefined behavior. For example `ftell` returns invalid values.

Comment (MiranDMC) [reply cleo_plugins/FileSystemOperations/FileUtils.cpp]: Updated.

**Review by x87 — APPROVED** [review]

**Review by x87 — APPROVED** [review]

**Pull request merged**